### PR TITLE
Added basic value providers in workload module

### PR DESCRIPTION
**Pull request by bistaastha**
- Added value-provider-interface and uniform-random-value-provider
- Added unit tests for value providers
- UniformRandomValueProvider test fix and reference value provider addition

Signed-off-by: Aastha Bist <abist119@gmail.com>

### CODE REVIEW

File: packages/caliper-core/lib/worker/workload/declarative/value-providers/value-provider-interface.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+class ValueProviderInterface {
+    /**
+     * Initialize an instance of ValueProviderInterface
+     * @param {object} options Options for configured parameter generation
+     * @param {object} variables Caliper provided variables
+     * @param {object} parameters User provided variables
+     * @param {object} valueProviderFactory ValueProviderFactory object reference
+     */
+    constructor(
+        options = {},
```
Comment (aklenik) [inline packages/caliper-core/lib/worker/workload/declarative/value-providers/value-provider-interface.js:28]: On second thought, we shouldn't assign default values to these parameters. If something is not passed, it should be undefined, and checked later by inheriting classes.

File: packages/caliper-core/lib/worker/workload/declarative/value-providers/value-provider-interface.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        variables = {},
+        parameters = {},
+        valueProviderFactory
+    ) {
+        this.options = options;
+        this.variables = variables;
+        this.parameters = parameters;
+        if (valueProviderFactory === undefined) {
```
Comment (aklenik) [inline packages/caliper-core/lib/worker/workload/declarative/value-providers/value-provider-interface.js:36]: This check should be left to inheriting classes, if they need the factory. We should remain flexible in the base class.

File: packages/caliper-core/lib/worker/workload/declarative/value-providers/value-provider-interface.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            throw new Error('valueProviderFactory is undefined');
+        }
+        this.valueProviderFactory = valueProviderFactory;
+    }
+    /**
+     * Generate values for corresponding options provided
+     */
```
Comment (aklenik) [inline packages/caliper-core/lib/worker/workload/declarative/value-providers/value-provider-interface.js:43]: The doc should contain a return statement (with type `object`). And the code should also contain a `return undefined;` after the `throw` clause, just to pass lint checks.

File: packages/caliper-core/lib/worker/workload/declarative/value-providers/variable-reference-value-provider.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    /**
+     * Initialize an instance of VariableReferenceValueProvider
+     * @param {object} options Options for configured parameter generation
+     * @param {object} variables Caliper provided variables
+     * @param {object} parameters User provided variables
+     * @param {object} valueProviderFactory ValueProviderFactory object reference
+     */
+    constructor(options, variables, parameters, valueProviderFactory) {
+        super(options, variables, parameters, valueProviderFactory);
+
```
Comment (aklenik) [inline packages/caliper-core/lib/worker/workload/declarative/value-providers/variable-reference-value-provider.js:32]: It should be checked whether `this.options` is undefined, and throw an error if it is.

File: packages/caliper-core/lib/worker/workload/declarative/value-providers/variable-reference-value-provider.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        if (
+            this.options.name === undefined ||
+            typeof this.options.name !== 'string'
+        ) {
+            throw new Error(
+                `Incorrect variable name: ${options.name}`
+            );
+        }
+
```
Comment (aklenik) [inline packages/caliper-core/lib/worker/workload/declarative/value-providers/variable-reference-value-provider.js:41]: It should be checked whether `this.variables` is undefined, and throw an error if it is.

File: packages/caliper-core/lib/worker/workload/declarative/value-providers/parameter-reference-value-provider.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    /**
+     * Initialize an instance of ParameterReferenceValueProvider
+     * @param {object} options Options for configured parameter generation
+     * @param {object} variables Caliper provided variables
+     * @param {object} parameters User provided variables
+     * @param {object} valueProviderFactory ValueProviderFactory object reference
+     */
+    constructor(options, variables, parameters, valueProviderFactory) {
+        super(options, variables, parameters, valueProviderFactory);
+
```
Comment (aklenik) [inline packages/caliper-core/lib/worker/workload/declarative/value-providers/parameter-reference-value-provider.js:32]: It should be checked whether `this.options` is undefined, and throw an error if it is.

**Review by aklenik — CHANGES_REQUESTED** [review]
A few more checks are still needed once the default parameter values are changed (back to undefined). Tests should be updated accordingly.

File: packages/caliper-core/lib/worker/workload/declarative/value-providers/value-provider-interface.js (hunk 3 above)
Comment (bistaastha) [reply packages/caliper-core/lib/worker/workload/declarative/value-providers/value-provider-interface.js:43]: Doing this fails the checks as there is unreachable return undefined after the throw is unreachable code.

Comment (aklenik) [reply packages/caliper-core/lib/worker/workload/declarative/value-providers/value-provider-interface.js:40]: @bistaastha I think you can disable that rule only for that line, with a remark stating the reason (e.g., documentation correctness is more important).

**Review by aklenik — APPROVED** [review]

**Pull request merged**